### PR TITLE
AIP-4801 - Integration Test for Email on Failure and OpsGenie Alerting

### PR DESCRIPTION
**Pull request by hsezhiyan**
In this PR, we add some functionality to the `raise_error_flow.py` test, which intentionally fails to ensure the Integration Testing suite itself is working.

In particular, we add `--notify` when launching the test so that when it fails, it sends an email to an OpsGenie email (`aip-workflow-sdk-email-notify-test@zillowgroup.opsgenie.net`). We then use the OpsGenie API to verify an OpsGenie alert was created, and then close that alert.

### CODE REVIEW

**Review by cloudw — APPROVED** [review]
Look good to me. 

Does other flow failure also result in ops genie being triggered? For example if bugs exist in my branch failing all the integration test, does that mean I need to close a bunch of alerts manually in OpsGenie, or are their alert level lower than P5 such that I can ignore?

**Comment by hsezhiyan** [comment]
> Look good to me.
> 
> Does other flow failure also result in ops genie being triggered? For example if bugs exist in my branch failing all the integration test, does that mean I need to close a bunch of alerts manually in OpsGenie, or are their alert level lower than P5 such that I can ignore?

Hey Yun,
Nope, only one flow will result in an OpsGenie alert being triggered (because we launch it with `--notify`). If an OpsGenie alert is triggered, it will be closed by the integration tests, so you'll ideally never have to manually close alerts. Also, all the alerts will be P5, so they're low priority anyways.

File: metaflow/plugins/kfp/tests/run_integration_tests.py [hunk 1]
```diff
@@ -100,19 +103,60 @@ def test_s3_sensor_flow(pytestconfig) -> None:
...
         test_cmd += (
             f"--no-s3-code-package --base-image {pytestconfig.getoption('image')}"
         )
 
-    exponential_backoff_from_platform_errors(test_cmd, 1)
+    kfp_run_id = exponential_backoff_from_platform_errors(test_cmd, 1)
+    # TODO use @secret (when it's released) to store this API key
+    opsgenie_auth_headers = {
+        "Content-Type": "application/json",
+        "Authorization": "GenieKey 21bda224-d719-42e5-aceb-35616eb26a6c",
```
Comment (alexlatchford) [inline metaflow/plugins/kfp/tests/run_integration_tests.py:126]: Can you please remove this from the repo. Having an API key in a public repo really isn't acceptable and we'll need to think through how best to store this unfortunately.

**Review by alexlatchford — CHANGES_REQUESTED** [review]

File: metaflow/plugins/kfp/tests/run_integration_tests.py (hunk 1 above)
Comment (alexlatchford) [reply metaflow/plugins/kfp/tests/run_integration_tests.py:126]: Can we store this in a Gitlab secret and when these tests are run via Gitlab we can pull this secret in via the CLI command in the `.gitlab-ci.yml`?

Comment (hsezhiyan) [reply metaflow/plugins/kfp/tests/run_integration_tests.py:126]: Got it, my bad! I just regenerated the key, so this key should no longer pose a threat.

Comment (hsezhiyan) [reply metaflow/plugins/kfp/tests/run_integration_tests.py]: @alexlatchford thanks for this quick catch! I updated this PR to use a Gitlab "masked" variable so it's not revealed in plaintext. I also updated to using a new API key.

File: metaflow/plugins/kfp/tests/run_integration_tests.py [hunk 2]
```diff
@@ -100,19 +103,60 @@ def test_s3_sensor_flow(pytestconfig) -> None:
...
+        f"--experiment metaflow_test --tag test_t1 --notify "
     )
     if pytestconfig.getoption("image"):
         test_cmd += (
             f"--no-s3-code-package --base-image {pytestconfig.getoption('image')}"
         )
 
-    exponential_backoff_from_platform_errors(test_cmd, 1)
+    kfp_run_id = exponential_backoff_from_platform_errors(test_cmd, 1)
+    # TODO use @secret (when it's released) to store this API key
```
Comment (alexlatchford) [inline metaflow/plugins/kfp/tests/run_integration_tests.py:123]: This is an interesting one, I think we've got some thinking about how we handle importing secrets given the current strategy DevEx has for storing and linking secrets. I'd actually say to remove the TODO as I think this approach is fine and maybe if/when we get to `@secret` if it makes sense to add this in.

**Review by alexlatchford — APPROVED** [review]

File: metaflow/plugins/kfp/tests/run_integration_tests.py (hunk 2 above)
Comment (hsezhiyan) [reply metaflow/plugins/kfp/tests/run_integration_tests.py]: Cool, removed the TODO. I forgot about this TODO and I don't think it's needed anymore.

File: metaflow/plugins/kfp/tests/run_integration_tests.py [hunk 3]
```diff
@@ -100,19 +103,59 @@ def test_s3_sensor_flow(pytestconfig) -> None:
     return
 
 
-# this test ensures the integration tests fail correctly
-def test_raise_failure_flow(pytestconfig) -> None:
+# This test ensures the integration tests fail correctly
+# and that when a test fails, an OpsGenie email is sent.
```
Comment (talebzeghmi) [inline metaflow/plugins/kfp/tests/run_integration_tests.py:107]: nit: more precise would be to say that it ensures the Flow fails correctly, and when the flow fails, an OpsGenie email is sent.

File: metaflow/plugins/kfp/tests/run_integration_tests.py [hunk 4]
```diff
@@ -240,3 +283,7 @@ def exponential_backoff_from_platform_errors(
         raise MetaflowException(
             "KFAM issues not resolved after successive backoff attempts."
         )
+    kfp_run_id = re.search("Metaflow run_id=(.*)\n", run_and_wait_process.stderr).group(
```
Comment (talebzeghmi) [inline metaflow/plugins/kfp/tests/run_integration_tests.py:286]: this is test code, but for production code, I highly recommend to stay away from regular expressions.

**Review by talebzeghmi — APPROVED** [review]

**Pull request merged**